tests/integration_tests/functional/test_serial_io.py

```diff
@@ -0,0 +1,159 @@
+    __test__ = False
+
+    def __init__(self, match_string):
+        self._string = match_string
+        self._input = ""
+
+    def match(self, input_char) -> bool:
+        """
+        Process one char at a time and build `_input` string.
+        Preserve built `_input` if partially matches `_string`.
+        Return True when `_input` is the same as `_string`.
+        """
+        self._input += str(input_char)
+        if self._input == self._string[:len(self._input)]:
+            if len(self._input) == len(self._string):
+                self._input = ""
+                return True
+            return False
+
+        self._input = self._input[1:]
+        return False
+
+
+class TestState(MatchStaticString):
+    """
+    Generic test state object
+    """
+    # Prevent state objects from being collected by pytest.
+    __test__ = False
+
+    def __init__(self, match_string=''):
+        MatchStaticString.__init__(self, match_string)
+        print('\n*** Current test state: ', str(self), end='')
+
+    def handle_input(self, microvm, input_char):
+        """
+        Handle input event and return next state
+        """
+
+    def __repr__(self):
+        """
+        Leverages the __str__ method to describe the State.
+        """
+        return self.__str__()
+
+    def __str__(self):
+        """
+        Return state name.
+        """
+        return self.__class__.__name__
+
+
+class WaitLogin(TestState):
+    """
+    Initial state when we wait for the login prompt.
+    """
+    def handle_input(self, microvm, input_char) -> TestState:
+        if self.match(input_char):
+            # Send login username.
+            microvm.serial_input("root")
+            return WaitPasswordPrompt("Password:")
+        return self
+
+
+class WaitPasswordPrompt(TestState):
+    """
+    Waiting for the password prompt to be shown.
+    """
+    def handle_input(self, microvm, input_char) -> TestState:
+        if self.match(input_char):
+            microvm.serial_input("root")
+            # Wait 1 second for shell
+            time.sleep(1)
+            microvm.serial_input("id")
+            return WaitIDResult("uid=0(root) gid=0(root) groups=0(root)")
+        return self
+
+
+class WaitIDResult(TestState):
+    """
+    Waiting for the console to show the result of the 'id' shell command.
+    """
+    def handle_input(self, microvm, input_char) -> TestState:
+        """
+        Handle input event and return next state
+        """
+        if self.match(input_char):
+            return TestFinished()
+        return self
+
+
+class TestFinished(TestState):
+    """
+    Test complete and successful.
+    """
+    def handle_input(self, microvm, input_char) -> TestState:
+        """
+        Return self since the test is about to end.
+        """
+        return self
+
+
+def test_serial_console_login(test_microvm_with_ssh_screen):
+    """Test serial console login."""
+    microvm = test_microvm_with_ssh_screen
```

Comment:
Why doesn t this work?
```
def test_serial_console_login(test_microvm_with_ssh):
    """Test serial console login."""
    microvm = test_microvm_with_ssh
    microvm.jailer.daemonize = False
...
...
...
```

Reply:
It works, but I like more the fixture approach. However your suggestion results in less code added.

Reply:
Fixtures are evil :speak_no_evil:, avoid wherever possible.